### PR TITLE
chore: update code to PHPStan level 2

### PR DESCRIPTION
**Pull request by carlos-granados**
Adds the necessary changes to update our codebase to PHPStan level 2

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,4 @@ behat.yml
 vendor
 composer.lock
 .phpunit.result.cache
+phpstan.neon
```
Comment (carlos-granados) [inline .gitignore:8]: Allows having a local phpstan.neon files that overrides some rules or modifies some settings

File: src/Behat/Behat/Context/Cli/ContextSnippetsController.php [hunk 2]
```diff
@@ -87,5 +87,6 @@ public function execute(InputInterface $input, OutputInterface $output)
                 new FixedPatternIdentifier($input->getOption('snippets-type'))
             ))
         );
+        return null;
```
Comment (carlos-granados) [inline src/Behat/Behat/Context/Cli/ContextSnippetsController.php:90]: These controller functions need to return a value

File: src/Behat/Behat/Definition/Pattern/Policy/RegexPatternPolicy.php [hunk 3]
```diff
@@ -22,7 +22,7 @@
 final class RegexPatternPolicy implements PatternPolicy
 {
     /**
-     * @var string[string]
+     * @var array<string,string>
```
Comment (carlos-granados) [inline src/Behat/Behat/Definition/Pattern/Policy/RegexPatternPolicy.php:25]: Annotation was wrong

File: src/Behat/Behat/EventDispatcher/Cli/StopOnFailureController.php [hunk 4]
```diff
@@ -81,12 +81,7 @@ public function configure(Command $command)
     }
 
     /**
-     * Executes controller.
-     *
-     * @param InputInterface  $input
-     * @param OutputInterface $output
-     *
-     * @return null|integer
+     * {@inheritdoc}
```
Comment (carlos-granados) [inline src/Behat/Behat/EventDispatcher/Cli/StopOnFailureController.php:84]: Prefer to use the base annotation for all controller classes

File: src/Behat/Behat/Gherkin/ServiceContainer/GherkinExtension.php [hunk 5]
```diff
@@ -79,25 +79,26 @@ public function initialize(ExtensionManager $extensionManager)
      */
     public function configure(ArrayNodeDefinition $builder)
     {
-        $builder
+        $childrenBuilder = $builder
```
Comment (carlos-granados) [inline src/Behat/Behat/Gherkin/ServiceContainer/GherkinExtension.php:82]: In some cases the `end()` function will return a type that cannot be used to chain other calls to other config functions because these other functions are not defined in the returned type. We split the chain of calls so that the functions are always defined in the returned types

File: src/Behat/Behat/HelperContainer/BuiltInServiceContainer.php [hunk 6]
```diff
@@ -115,13 +115,8 @@ private function getAndValidateServiceSchema($id)
...
     /**
      * Gets and validates a class from schema.
-     *
-     * @param string       $id
-     * @param string|array $schema
-     *
-     * @return string
      */
-    private function getAndValidateClass($id, array $schema)
+    private function getAndValidateClass(string $id, array $schema): string
```
Comment (carlos-granados) [inline src/Behat/Behat/HelperContainer/BuiltInServiceContainer.php:119]: The annotation for the $schema parameter was wrong. I changed this to use proper php types everywhere

File: src/Behat/Behat/Output/Node/EventListener/AST/ScenarioNodeListener.php [hunk 7]
```diff
@@ -83,9 +83,9 @@ public function listenEvent(Formatter $formatter, Event $event, $eventName)
      * @param ScenarioLikeTested|AfterSetup $event
      * @param string                        $eventName
      */
-    private function printHeaderOnBeforeEvent(Formatter $formatter, ScenarioLikeTested $event, $eventName)
+    private function printHeaderOnBeforeEvent(Formatter $formatter, ScenarioLikeTested|AfterSetup $event, $eventName)
```
Comment (carlos-granados) [inline src/Behat/Behat/Output/Node/EventListener/AST/ScenarioNodeListener.php:86]: The event can implement any of these interfaces

File: src/Behat/Behat/Output/Node/EventListener/AST/ScenarioNodeListener.php [hunk 8]
```diff
@@ -83,9 +83,9 @@ public function listenEvent(Formatter $formatter, Event $event, $eventName)
...
     {
-        if ($this->beforeEventName !== $eventName || !$event instanceof AfterSetup) {
+        if ($this->beforeEventName !== $eventName || !$event instanceof AfterSetup || !$event instanceof ScenarioLikeTested) {
```
Comment (carlos-granados) [inline src/Behat/Behat/Output/Node/EventListener/AST/ScenarioNodeListener.php:88]: But we only want to deal with it when it implements both

File: src/Behat/Behat/Output/Node/EventListener/Statistics/HookStatsListener.php [hunk 9]
```diff
@@ -111,6 +112,7 @@ private function captureAfterHookStats(HookedTeardown $teardown)
     private function captureHookStat(CallResult $hookCallResult)
     {
         $call = $hookCallResult->getCall();
+        assert($call instanceof HookCall);
```
Comment (carlos-granados) [inline src/Behat/Behat/Output/Node/EventListener/Statistics/HookStatsListener.php:115]: type narrowing

File: src/Behat/Behat/Output/Node/Printer/JUnit/JUnitSetupPrinter.php [hunk 10]
```diff
@@ -52,23 +54,20 @@ public function printTeardown(Formatter $formatter, Teardown $teardown)
         }
     }
 
-    /**
```
Comment (carlos-granados) [inline src/Behat/Behat/Output/Node/Printer/JUnit/JUnitSetupPrinter.php:55]: Unnecessary annotations

File: src/Behat/Behat/Output/Node/Printer/JUnit/JUnitSetupPrinter.php [hunk 11]
```diff
@@ -52,23 +54,20 @@ public function printTeardown(Formatter $formatter, Teardown $teardown)
...
-     * @param Formatter $formatter
-     * @param CallResults $results
-     * @param string $messageType
-     */
-    private function handleHookCalls(Formatter $formatter, CallResults $results, $messageType)
+    private function handleHookCalls(Formatter $formatter, CallResults $results, string $messageType): void
     {
-        /** @var CallResult $hookCallResult */
         foreach ($results as $hookCallResult) {
+            assert($hookCallResult instanceof CallResult);
```
Comment (carlos-granados) [inline src/Behat/Behat/Output/Node/Printer/JUnit/JUnitSetupPrinter.php:60]: Prefer to use assertions to @var annotations

File: src/Behat/Behat/Snippet/SnippetRegistry.php [hunk 12]
```diff
@@ -55,7 +55,7 @@ public function registerSnippetGenerator(SnippetGenerator $generator)
      * @param Environment $environment
      * @param StepNode    $step
      *
-     * @return null|Snippet
+     * @return void
```
Comment (carlos-granados) [inline src/Behat/Behat/Snippet/SnippetRegistry.php:58]: Function does not return anything

File: src/Behat/Behat/Tester/ServiceContainer/TesterExtension.php [hunk 13]
```diff
@@ -81,8 +81,6 @@ public function configure(ArrayNodeDefinition $builder)
                             ? sys_get_temp_dir() . DIRECTORY_SEPARATOR . 'behat_rerun_cache'
                             : null
                     )
-                ->end()
```
Comment (carlos-granados) [inline src/Behat/Behat/Tester/ServiceContainer/TesterExtension.php:84]: These `end()` calls are not really needed because they just return a value which is not used and they return a type which cannot be chained with another end call, so I just removed them

File: src/Behat/Hook/Hook.php [hunk 14]
```diff
@@ -18,4 +18,5 @@
  */
 interface Hook
 {
+    public function getFilterString(): ?string;
```
Comment (carlos-granados) [inline src/Behat/Hook/Hook.php:21]: Adding this function call is a BC break but I think we can allow it. Otherwise we cannot guarantee that a class implementing this interface will have a filterString property. I don't expect anyone to have defined other classes that implement this interface which is applied to Hook attributes

File: src/Behat/Step/Definition.php [hunk 15]
```diff
@@ -18,4 +18,5 @@
  */
 interface Definition
 {
+    public function getPattern(): ?string;
```
Comment (carlos-granados) [inline src/Behat/Step/Definition.php:21]: Same here for access to the pattern property, again it is a BC break but I don't expect anyone to have defined other classes that implement this interface

File: src/Behat/Testwork/Specification/SpecificationIterator.php [hunk 16]
```diff
@@ -18,7 +18,7 @@
  *
  * @author Konstantin Kudryashov <ever.zet@gmail.com>
  *
- * @extends Iterator<int, Suite>
+ * @extends Iterator<int, mixed>
```
Comment (carlos-granados) [inline src/Behat/Testwork/Specification/SpecificationIterator.php:21]: This annotation was wrong, my mistake, looking at its definition I thought it would implement an iterator of Suites but it actually implements an iterator of specifications which are mixed

File: .gitignore (hunk 1 above)
Comment (stof) [reply .gitignore:8]: I always failed to see the use case for that. Running phpstan locally should use the same config than the CI IMO

File: src/Behat/Behat/EventDispatcher/Cli/StopOnFailureController.php (hunk 4 above)
Comment (stof) [reply src/Behat/Behat/EventDispatcher/Cli/StopOnFailureController.php:84]: The inline tag `{@inheritdoc}` is about inheriting the parent description, which is useful only to merge it with more description. It has nothing to do with inheriting tags like `@param` (which are inherited automatically anyway in phpdoc tools).
You should either remove the phpdoc entirely or use the `@inheritdoc` tag as an explicit marker meaning "the phpdoc is inherited from the parent, I did not forgot to add it".
My preference is to remove it entirely.

File: src/Behat/Behat/Output/Node/EventListener/AST/ScenarioNodeListener.php (hunk 7 above)
Comment (stof) [reply src/Behat/Behat/Output/Node/EventListener/AST/ScenarioNodeListener.php:86]: no need to change this native type. The check in `listenEvent` ensures that we always get a ScenarioLikeTested

File: src/Behat/Behat/Output/Node/EventListener/AST/ScenarioNodeListener.php [hunk 17]
```diff
@@ -103,7 +103,7 @@ private function printHeaderOnBeforeEvent(Formatter $formatter, ScenarioLikeTest
      * @param ScenarioLikeTested|AfterTested $event
      * @param string                         $eventName
      */
-    private function printFooterOnAfterEvent(Formatter $formatter, ScenarioLikeTested $event, $eventName)
+    private function printFooterOnAfterEvent(Formatter $formatter, ScenarioLikeTested|AfterTested $event, $eventName)
```
Comment (stof) [inline src/Behat/Behat/Output/Node/EventListener/AST/ScenarioNodeListener.php:106]: same here

File: .gitignore (hunk 1 above)
Comment (carlos-granados) [reply .gitignore:8]: I have this configuration in this file:
```
includes:
	- phpstan.dist.neon

parameters:
	editorUrl: 'phpstorm://open?file=%%file%%&line=%%line%%'
```
This allows me to open phpstan found errors in phpstorm, but obviously this config is local to me

File: src/Behat/Behat/Output/Node/Printer/JUnit/JUnitSetupPrinter.php (hunk 11 above)
Comment (stof) [reply src/Behat/Behat/Output/Node/Printer/JUnit/JUnitSetupPrinter.php:60]: The right solution here is to add `@template-implements \IteratorAggregate<array_key, CallResult>` on the `CallResults` class, removing the need for local type refinements.

I agree that assertions are better than `@var` because wrong assertions will make our CI fail (assuming we configured GHA to use the development ini config in setup-php), but proper type inference is even better from a type safety point of view, as it is better to fail during type checking than at runtime.

File: .gitignore (hunk 1 above)
Comment (stof) [reply .gitignore:8]: ah, I did not know that phpstan had this parameter. This is indeed a valid use case.

File: src/Behat/Testwork/Specification/SpecificationIterator.php (hunk 16 above)
Comment (stof) [reply src/Behat/Testwork/Specification/SpecificationIterator.php:21]: It should actually be generic instead of using `mixed`, allowing better types for implementations of this interface. For instance, `LazyFeatureIterator` is an iterator over `FeatureNode`, not over `mixed`.
Basically, it should forward the (covariant) template type from `Iterator` for the value type.

`NoSpecificationsIterator` will be a `SpecificationIterator<never>` and `SpecificationArrayIterator` will be generic (inferred from the constructor argument passing the array). `GroupedSpecificationIterator` will also be generic (and its static `group` method as well).

Comment (carlos-granados) [reply src/Behat/Testwork/Specification/SpecificationIterator.php:21]: You lost me here. What would be your proposed solution?

File: src/Behat/Behat/EventDispatcher/Cli/StopOnFailureController.php (hunk 4 above)
Comment (carlos-granados) [reply src/Behat/Behat/EventDispatcher/Cli/StopOnFailureController.php:84]: OK, understood, I also prefer to remove it

File: src/Behat/Behat/Output/Node/EventListener/AST/ScenarioNodeListener.php (hunk 7 above)
Comment (carlos-granados) [reply src/Behat/Behat/Output/Node/EventListener/AST/ScenarioNodeListener.php:86]: Yeah, my solution was wrong, I actually needed to fix the doc-block annotation but in fact it is even better to just remove it

File: src/Behat/Testwork/Specification/SpecificationIterator.php (hunk 16 above)
Comment (stof) [reply src/Behat/Testwork/Specification/SpecificationIterator.php:21]:
```php
/**
 * @template-covariant T
 * @extends Iterator<int, T>
 */
```

And then updating child classes to define the type of specification they work with.

File: src/Behat/Behat/Output/Node/Printer/JUnit/JUnitSetupPrinter.php (hunk 11 above)
Comment (carlos-granados) [reply src/Behat/Behat/Output/Node/Printer/JUnit/JUnitSetupPrinter.php:60]: The `CallResults` class already had the `@implements` annotation so this assert was not actually needed (and neither a couple of other asserts). I had converted the @var annotations to asserts without checking if they were actually needed, now they have been removed

File: src/Behat/Testwork/Specification/SpecificationIterator.php (hunk 16 above)
Comment (carlos-granados) [reply src/Behat/Testwork/Specification/SpecificationIterator.php:21]: This surpasses my knowledge/understanding of generics, I am in a bit of a loss to what would actually need to go in the child classes, if you can provide some code samples (or a new commit), I would be grateful

Comment (stof) [reply src/Behat/Testwork/Specification/SpecificationIterator.php:21]: I started adding generic types for `SpecificationIterator`, and this is quite far-reaching (which is mostly related to the fact that `Behat/Testwork` is written to be agnostic of Gherkin and so does not hardcode that the specification being used is FeatureNode). I'll submit that as a separate PR instead.

Comment (stof) [reply src/Behat/Testwork/Specification/SpecificationIterator.php:21]: I submitted it as #1575

File: .gitignore (hunk 1 above)
Comment (acoulton) [reply .gitignore:8]: Very useful tip, I didn't know that either - thanks!

File: src/Behat/Behat/Gherkin/ServiceContainer/GherkinExtension.php (hunk 5 above)
Comment (acoulton) [reply src/Behat/Behat/Gherkin/ServiceContainer/GherkinExtension.php:82]: IMO this is also more readable so 👍

File: src/Behat/Behat/Output/Node/Printer/JUnit/JUnitSetupPrinter.php [hunk 18]
```diff
@@ -52,23 +54,17 @@ public function printTeardown(Formatter $formatter, Teardown $teardown)
...
-                /** @var HookCall $call */
                 $call = $hookCallResult->getCall();
                 $scope = $call->getScope();
-                /** @var JUnitOutputPrinter $outputPrinter */
                 $outputPrinter = $formatter->getOutputPrinter();
+                assert($outputPrinter instanceof JUnitOutputPrinter);
 
-                $message = $call->getCallee()->getName();
+                $callee = $call->getCallee();
+                $message = $callee->getName();
```
Comment (acoulton) [inline src/Behat/Behat/Output/Node/Printer/JUnit/JUnitSetupPrinter.php:67]: Not sure why this change?

File: src/Behat/Hook/Hook.php (hunk 14 above)
Comment (acoulton) [reply src/Behat/Hook/Hook.php:21]: We should remember to mention this (and Definition) in the changelog. If we hit similar things in future we should probably pull to a separate PR for visibility, but I think it's OK here.

**Review by acoulton — APPROVED** [review]
Thanks @carlos-granados :)
